File: two_pointers/trapping_rain_water/trap_rain_water.go
// Package trapwater will solve the trap_rain_water question.
package trapwater

func trapWater(heights []int) int {
	if len(heights) == 0 {
		return 0
	}
	result := 0
	l, r := 0, len(heights)-1
	lMax, rMax := heights[l], heights[r]
	for l < r {
		if lMax < rMax {
			l += 1
			lMax = max(lMax, heights[l])
			result += lMax - heights[l]
			continue
		}
		r -= 1
		rMax = max(rMax, heights[r])
		result += rMax - heights[r]

	}

	return result
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: two_pointers/trapping_rain_water/trap_rain_water.py
def trap(heights):
    if not heights:
        return 0

    left_pointer, right_pointer = 0, len(heights) - 1
    left_max, right_max = heights[left_pointer], heights[right_pointer]
    result = 0
    while left_pointer < right_pointer:
        if left_max < right_max:
            left_pointer += 1
            left_max = max(left_max, heights[left_pointer])
            # Do not worry about negative numbers because we set left_max first.
            result += left_max - heights[left_pointer]
        else:
            right_pointer -= 1
            right_max = max(right_max, heights[right_pointer])
            # Do not worry about negative numbers because we set right_max first.
            result += right_max - heights[right_pointer]
    return result
